exchange/src/exchange/tickers/engine/level_update_generator.cpp

```diff
@@ -1,57 +1,35 @@
 #include "level_update_generator.hpp"
 
-#include <unordered_set>
-
 namespace nutc {
 namespace matching {
-
-// There is a more efficient way to do this, but it's O(levels) which will be <=100
-// in most cases and only called on tick
-// so simplicity > performance
 std::vector<ob_update>
-get_updates(const std::string& ticker, const OrderBook& before, const OrderBook& after)
+LevelUpdateGenerator::get_updates(const std::string& ticker)
 {
-    std::unordered_set<double> buy_levels{};
-    std::unordered_set<double> ask_levels{};
-    for (auto [price, _] : before.get_levels(util::Side::buy)) {
-        buy_levels.insert(price);
-    }
-    for (auto [price, _] : before.get_levels(util::Side::sell)) {
-        ask_levels.insert(price);
-    }
-    for (auto [price, _] : after.get_levels(util::Side::buy)) {
-        buy_levels.insert(price);
+    std::vector<ob_update> updates;
+    updates.reserve(updated_buy_levels_.size() + updated_sell_levels_.size());
+
+    for (auto [price, quantity] : updated_buy_levels_) {
+        updates.emplace_back(ticker, util::Side::buy, price, quantity);
     }
-    for (auto [price, _] : after.get_levels(util::Side::sell)) {
-        ask_levels.insert(price);
+    for (auto [price, quantity] : updated_sell_levels_) {
+        updates.emplace_back(ticker, util::Side::sell, price, quantity);
     }
 
-    std::vector<ob_update> updates{};
-    for (auto price : buy_levels) {
-        double before_quantity = before.get_level(util::Side::buy, price);
-        double after_quantity = after.get_level(util::Side::buy, price);
-
-        if (before_quantity == after_quantity)
-            continue;
-
-        // Round to 3 places
-        before_quantity = std::round(before_quantity * 1000) / 1000;
-        after_quantity = std::round(after_quantity * 1000) / 1000;
+    return updates;
+}
 
-        updates.emplace_back(ticker, util::Side::buy, price, after_quantity);
+void
+LevelUpdateGenerator::record_level_change(
+    util::Side side, double price, double new_quantity
+)
+{
+    assert(std::round(price * 100) / 100 == price);
```

Comment:
you may hit FPE bugs here

Reply:
good point, but this is checking to make sure we don't :)
This is asserted/handled elsewhere but this is just an extra sanity check, not really required

In more detail: std::round(price*100)/100 should be deterministic, i.e. 5.000001 and 5.00000002 should round to the exact same number